javalib/src/main/scala/java/util/ArrayDeque.scala

```diff
@@ -171,55 +193,175 @@ class ArrayDeque[E] private (private val inner: js.Array[E])
+            _hasNext = false
+          } else {
+            nextIndex -= 1
+            if (nextIndex < 0)
+              nextIndex = inner.length - 1
+          }
+        }
+
+        inner(lastIndex)
       }
 
       override def remove(): Unit = {
         checkStatus()
-        if (index < 0 || index >= inner.length) {
+        if (lastIndex == -1) {
           throw new IllegalStateException()
         } else {
-          arrayRemove(inner, index)
+          val laterShifted = removeAt(lastIndex)
+          expectedStatus = self.status
+          if (forward && laterShifted) {
+            nextIndex -= 1
+          }
+
+          if (!forward && !laterShifted) {
+            nextIndex += 1
+          }
         }
       }
     }
   }
 
-  def iterator(): Iterator[E] =
-    failFastIterator(-1, x => (x + 1))
+  def iterator(): Iterator[E] = failFastIterator(true)
 
-  def descendingIterator(): Iterator[E] =
-    failFastIterator(inner.length, x => (x - 1))
+  def descendingIterator(): Iterator[E] = failFastIterator(false)
 
-  override def contains(o: Any): Boolean =
-    arrayExists(inner)(Objects.equals(_, o))
+  override def contains(o: Any): Boolean = firstIndexOf(o) != -1
 
   override def remove(o: Any): Boolean = removeFirstOccurrence(o)
 
   override def clear(): Unit = {
     if (!isEmpty())
       status += 1
-    inner.length = 0
+    empty = true
+    startIndex = 0
+    endIndex = inner.length
+  }
+
+  private def firstIndexOf(o: Any): Int = {
+    // scalastyle:off return
+    if (isEmpty())
+      return -1
+    val inner = this.inner // local copy
+    val capacity = inner.length // local copy
+    val endIndex = this.endIndex // local copy
+    var i = startIndex
+    do {
+      if (i >= capacity)
+        i = 0
+      if (Objects.equals(inner(i), o)) {
+        return i
+      }
+      i += 1 // let i overrun so we catch endIndex == capacity
+    } while (i != endIndex)
+    -1
+    // scalastyle:on return
+  }
+
+  private def ensureCapacityForAdd(): Unit = {
+    if (isEmpty()) {
+      // Nothing to do (constructor ensures capacity is always non-zero).
+    } else if (startIndex == 0 && endIndex == inner.length) {
+      inner.length *= 2
+      // no copying required: We just keep adding to the end.
+    } else if (startIndex == endIndex) {
+      val oldCapacity = inner.length
+      inner.length *= 2
+      // move beginning of array to end
+      var i = 0
+      while (i != endIndex) {
+        inner(i + oldCapacity) = inner(i)
+        i += 1
+      }
+      endIndex += oldCapacity
+    }
+  }
+
+  /* Removes the element at index [[target]]
+   *
+   * @returns true if elements after the target index were shifted.
+   * Returns false, if elements before the target index were shifted.
+   */
+  private def removeAt(target: Int): Boolean = {
+    if (target == startIndex) {
+      pollFirst()
+      false
+    } else if (target == endIndex - 1) {
+      pollLast()
+      true
```

Comment:
I have a hard time understanding what this result means. In this case and the case above, no element was shifted. Why does this one return `true` but the above returns `false`? How that translate to the behavior of the iterator?

Reply:
This should be clearer now: tl;dr if the result value here is such that it would trigger special behavior in the iterator, the iterator is done (so `nextIndex == -1`).

Reply:
Indeed, this is much clearer now, thanks.